rest/database_init_manager.go

```diff
@@ -0,0 +1,312 @@
+// is that evaluation of whether a database has been initialized is relatively inexpensive and is the responsibility
+// of the code in databaseInitWork.Run.
+type DatabaseInitManager struct {
+
+	// Set of active DatabaseInitWorkers.  Workers are removed from the set on completion.
+	workers     map[string]*DatabaseInitWorker
+	workersLock sync.Mutex
+
+	// collectionCompleteCallback is defined for testability only.
+	// Invoked after collection initialization is complete for each collection
+	collectionCompleteCallback collectionCallbackFunc
+
+	// databaseCompleteCallback is defined for testability only.
+	// Invoked after worker completes, but before worker is removed from workers set
+	databaseCompleteCallback func(databaseName string) // Callback for testability only
+}
+
+type collectionCallbackFunc func(dbName, collectionName string)
+
+// CollectionInitData defines the set of collections being created (by ScopeAneCollectionName), and the set of
+// indexes required for each collection.
+type CollectionInitData map[base.ScopeAndCollectionName]db.CollectionIndexesType
+
+// Initializes the database.  Will establish a new cluster connection using the provided server config.  Establishes a new
+// cluster-only N1QLStore based on the startup config to perform initialization.
+func (m *DatabaseInitManager) InitializeDatabase(ctx context.Context, startupConfig *StartupConfig, dbConfig *DatabaseConfig) (doneChan chan error, err error) {
+	m.workersLock.Lock()
+	defer m.workersLock.Unlock()
+	if m.workers == nil {
+		m.workers = make(map[string]*DatabaseInitWorker)
+	}
+	base.InfofCtx(ctx, base.KeyAll, "Initializing database %s ...",
+		base.MD(dbConfig.Name))
+	dbInitWorker, ok := m.workers[dbConfig.Name]
+	collectionSet := m.buildCollectionIndexData(dbConfig)
+	if ok {
+		// If worker exists for the database and the collection sets match, add watcher to the existing worker
+		if dbInitWorker.collectionsEqual(collectionSet) {
+			base.InfofCtx(ctx, base.KeyAll, "Found existing database initialization for database %s ...",
+				base.MD(dbConfig.Name))
+			doneChan, err := dbInitWorker.addWatcher()
+			return doneChan, err
+		}
+		// For a mismatch in collections, stop and remove the existing worker, then continue through to creation of new worker
+		dbInitWorker.Stop()
+		delete(m.workers, dbConfig.Name)
+	}
+
+	base.InfofCtx(ctx, base.KeyAll, "Starting new async initialization for database %s ...",
+		base.MD(dbConfig.Name))
+	couchbaseCluster, err := CreateCouchbaseClusterFromStartupConfig(startupConfig, base.PerUseClusterConnections)
+	if err != nil {
+		return nil, err
+	}
+
+	bucketName := dbConfig.Name
+	if dbConfig.Bucket != nil {
+		bucketName = *dbConfig.Bucket
+	}
+
+	// Initialize ClusterN1QLStore for the bucket.  Scope and collection name are set
+	// per-operation
+	n1qlStore, err := couchbaseCluster.GetClusterN1QLStore(bucketName, "", "")
```

Comment:
If these are set per operation, does it make sense to not pass scope and collection into the cluster and require it later only?

These are going to get separate agents anyway, it's almost like ClusterN1QLStore might be a misnomer until we can query system:indexes without a Scope based query.

Reply:
It should be one agent per scope, I agree.